### PR TITLE
Feat: Authentication for Function apps

### PR DESCRIPTION
**Pull request by RasmusGraabaek**
## Description

- [x] Implement middleware for User authentication in Function App

Part of: https://app.zenhub.com/workspaces/mosaic-60a6105157304f00119be86e/issues/gh/energinet-datahub/team-mosaic/182

## Quality

- [x] Documentation is updated
- [x] Release notes are updated
- [x] Package version is updated
- [x] Public types and methods are documented
- [x] Tests are implemented and executed locally

### CODE REVIEW

File: source/App/source/FunctionApp/Middleware/UserMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    // DO NOT inject scoped services in the middleware constructor.
+    // DO use scoped services in middleware by retrieving them from 'FunctionContext.InstanceServices'
+    // DO NOT store scoped services in fields or properties of the middleware object. See https://github.com/Azure/azure-functions-dotnet-worker/issues/1327#issuecomment-1434408603
+    public UserMiddleware()
+    {
+    }
+
+    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+    {
+        var httpRequestData = await context.GetHttpRequestDataAsync().ConfigureAwait(false);
```
Comment (RasmusGraabaek) [inline source/App/source/FunctionApp/Middleware/UserMiddleware.cs:46]:
```suggestion
        var httpRequestData = await context.GetHttpRequestDataAsync().ConfigureAwait(false);
```
What happens if "httpRequestData" is null?
Would it be preferable to call: next(context).ConfigureAwait(false) or throw an exception?

File: source/App/source/FunctionApp/Middleware/UserMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        if (httpRequestData != null)
+        {
+            var isUserSet = await CanSetUserAsync(context, httpRequestData).ConfigureAwait(false);
+            if (isUserSet)
+            {
+                // Next middleware
```
Comment (RasmusGraabaek) [inline source/App/source/FunctionApp/Middleware/UserMiddleware.cs:52]: sadsadsadsa

File: source/App/source/FunctionApp/Middleware/UserMiddleware.cs (hunk 1 above)
Comment (dstenroejl) [reply source/App/source/FunctionApp/Middleware/UserMiddleware.cs:46]: We decided to throw an exception

Comment (dstenroejl) [reply source/App/source/FunctionApp/Middleware/UserMiddleware.cs]: Fixed

File: source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.Core.App.FunctionApp.Extensions.Options;
+
+public class UserAuthenticationOptions
```
Comment (dstenroejl) [inline source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs:17]: Ideally this should not be a new type, but instead we should move the one we have in Web App to Common.

But let's talk to other developers as well and decide what is best.

File: source/App/documents/registrations/authorization.md [hunk 4]
```diff
@@ -10,7 +10,25 @@
...
 - DO create a `TUser` implementation that is convenient for your subsystem.
 
-## Authorization in Web Apps
+## Overview
+
+- Implementation
+    - [Azure Functions App](#azure-functions-app)
+    - [ASP.NET Core Web API](#aspnet-core-web-api)
+
+## Azure Functions App
```
Comment (dstenroejl) [inline source/App/documents/registrations/authorization.md:19]: Do we need to write more at the moment. The quick start example shows it all.

File: source/App/documents/release-notes/release-notes.md [hunk 5]
```diff
@@ -1,5 +1,13 @@
 # App Release notes
 
+## Version 12.1.0
+
+- In `FunctionApp` project:
```
Comment (dstenroejl) [inline source/App/documents/release-notes/release-notes.md:5]: Depending on what we decide with the options class, we might need to update this section.

File: source/App/documents/documentation.md [hunk 6]
```diff
@@ -50,13 +51,20 @@ Preparing an Azure Function App project:
 
    ```cs
    var host = new HostBuilder()
-       .ConfigureFunctionsWorkerDefaults()
+       .ConfigureFunctionsWorkerDefaults(worker =>
+       {
+           // Http => Authentication
+           worker.UseUserMiddlewareForIsolatedWorker<FrontendUser>();
```
Comment (defectiveAi) [inline source/App/documents/documentation.md:57]: This should be <SubsystemUser>

File: source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs (hunk 3 above)
Comment (defectiveAi) [reply source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs:17]: "IF we only want to enabled UserMiddleware (authentication) and not authorization, we actually do not need this class yet"

I do not understand - this class is needed and used for authentication only? All this stuff validates the token, telling who the user is.

File: source/App/documents/documentation.md (hunk 6 above)
Comment (defectiveAi) [reply source/App/documents/documentation.md:57]: Or maybe call it `TUser` - it is not clear where SubsystemUser comes from.

Comment (RasmusGraabaek) [reply source/App/documents/documentation.md:57]: Calling it <TUser>

File: source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs (hunk 3 above)
Comment (dstenroejl) [reply source/App/source/FunctionApp/Extensions/Options/UserAuthenticationOptions.cs:17]: Agree, my mistake, I should not have written authorization. What I mean is that the UserMiddleware doesn't need these options, so currently we don't need them.

File: source/App/documents/release-notes/release-notes.md (hunk 5 above)
Comment (dstenroejl) [reply source/App/documents/release-notes/release-notes.md:5]: We decided to not include a copy of the options class at the moment.

File: source/App/documents/registrations/authorization.md (hunk 4 above)
Comment (dstenroejl) [reply source/App/documents/registrations/authorization.md:19]: We decided it was OK for now.

**Review by dstenroejl — APPROVED** [review]
LGTM

**Pull request merged**